flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.streaming.api.functions.windowing.RichAllWindowFunction
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.functions.Accumulator
+
+import java.lang.Iterable
+import org.apache.flink.table.functions.AggregateFunction
+
+/**
+  * Computes the final aggregate value from incrementally computed aggreagtes.
+  *
+  * @param aggregates The aggregates to be computed
+  * @param aggFields the fields on which to apply the aggregate.
+  * @param forwardedFieldCount The fields to be carried from current row.
+  */
+class DataStreamProcTimeAggregateGlobalWindowFunction[W <: Window](
```

Comment:
@sunjincheng121 @fhueske 
Thanks for posting. I saw the remark also in the other post. I personally do not believe the processingFunction can be used in this case because of several reasons:
1) if we would use distinct, we cannot extend the implementation - so the design is not extensible/ scalable 
2) Looking at the interface/ example (https://ci.apache.org/projects/flink/flink-docs-release-1.2/dev/stream/process_function.html) i do not see how you can do eviction. We need to take into account the fact that for both JIRA (5653 and 5654) the windows are bounded. That means you need to also eliminate - in one based on time in the other based on number of elements. I am not sure how the processing function can support the eviction (do you have any examples for this)?
3) in case there is not support for eviction that means we need to work with some sort of states in order to keep the elements and manually implement the eviction logic. This would bring a lot of overhead and duplication of code just to not use the windows which directly offer this
4) we need to consider SQL as a whole (IMHO) - which means that we should consider various other parameters that can be added to the aggregation (for example LIMIT x / TOP y... - see https://issues.apache.org/jira/browse/FLINK-6075?filter=-2). This would require that before you apply the logic you have access to the elements to select the ones that are of interest. With processing function you cannot do this. On the other hand building this logic with windows allows you to extend the implementation to support various SQL features